server/pbench/bin/index-pbench

```diff
@@ -187,99 +319,303 @@ def es_index(es, actions, dbg=0):
+            # to their proper fields for each identifier. Now we can yield
+            # record for the identifiers.
+            for key,source in datum.items():
+                yield source
+        return
 
-def mk_metadata(fname, tb, mdconf, md5sum):
-    """Return a dict with metadata about a tarball"""
-    mtime = datetime.fromtimestamp(os.stat(fname)[stat.ST_MTIME])
+    def _make_source_individual(self):
+        """Read .csv files individually, emitting records for each row and
+        column coordinate."""
+        pass
 
-    mddict = {'generated-by': _NAME_,
-              'generated-by-version': _VERSION_,
-              'pbench-agent-version': mdconf.get("pbench", "rpm-version"),
-              'file-date': mtime.strftime("%Y-%m-%d"),
-              'file-name': fname,
-              'md5': md5sum
-    }
-    return mddict
+    def make_source(self):
+        """Simple jump method to pick the write source generator based on the
+        handler's prospectus."""
+        if not self.csv_files:
+            # If we do not have any csv files for this tool, ignore it.
+            return
+        if self.handler['@prospectus']['handling'] == 'unify':
+            gen = self._make_source_unified()
+        elif self.handler['@prospectus']['handling'] == 'individual':
+            gen = self._make_source_individual()
+        else:
+            raise Exception("Logic bomb!")
+        return gen
 
+    @staticmethod
+    def get_csv_files(handler, iteration, sample, host, tool, mdlog, tb):
+        """
+        Fetch the list of .csv files for this tool, fetch their headers, and
+        return a dictionary mapping their column headers to their field names. 
+        """
+        # FIXME: We need to be able to handle other tool groups
+        path = os.path.join(iteration, sample, "tools-default", host, tool, "csv")
+        paths = [x for x in tb.getnames() if x.find(path) >= 0 and tb.getmember(x).isfile()]
+        datafiles = []
+        for p in paths:
+            fname = os.path.basename(p)
+            try:
+                handler_rec = handler[fname]
+            except KeyError:
+                # Try an alias
+                alias_name = fname
+                try:
+                    fname = _aliases[alias_name]
+                    handler_rec = handler[fname]
+                except KeyError:
+                    # Ignore .csv files for which we don't have a handler
+                    #print("WARNING: no .csv handler for %s\n" % (p,), file=sys.stdout)
+                    continue
+            datafile = { "path": p, "basename": fname, 'handler': handler_rec }
+            # FIXME: we might have to deal with UTF-8 decoding issues
+            # reader = csv.reader(codecs.iterdecode(tb.extractfile(dfilepath), 'utf-8'))
+            datafile['reader'] = reader = csv.reader(open(os.path.join(_tmpdir, p)))
+            # FIXME: how should we handle .csv files that are empty?
```

Comment:
The change would look like:
reader = csv.reader(open(os.path.join(_tmpdir, p)))
if next(reader, None ):
        datafile['reader'] = reader
        datafile['header'] = next(reader)
        datafiles.append(datafile)